apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,509 @@
+<?php
+
+namespace Drupal\particle\Plugin\Preprocessor;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\particle\Tools\ParticleTools;
+
+/**
+ * Provides a theme preprocessor manager for Class based theme preprocessors.
+ *
+ * Drupal core does not provide built-in plugin support for theme files.
+ * The plugin discovery might before the front end theme is not active.
+ *
+ * The approach here is to use Drupal's State API to handle Preprocessor Class.
+ * New suggestions are read on page load and added to State API via STATE_ID.
+ * The Preprocessor Class Map is stored in cache and refreshed on change.
+ *
+ */
+class PreprocessorManager {
+
+  /**
+   * The State API ID for this Manager.
+   *
+   * @var string
+   */
+  const STATE_ID = 'preprocessor.map';
```

Comment:
This seems a little tricky if trying to support multiple Drupal themes, or use base themes. The state() for each of them would clobber each other unless the developer knows to manually prefix this with the theme name or whatever. I can get that maybe "support multiple themes" would be a follow up issue, but it seems like at least we should try to not cause bugs for multiple themes, even if they aren't necessarily supported.

Probably not a problem for most sites since they wouldn't be using multiple themes on the same site - it would be a multisite scenario in which case Drupal would probably only see the one that exists in the theme that's active on that site. But still just something to note/be aware of.

Maybe it should just be prefixed by `ParticleTools::getThemeName()`?

Reply:
This is a great suggestion!

Reply:
@mikecrittenden A class constant cannot contain an expression. So workaround here would be to:
* Prefix the theme name in the implementation of the State API (easiest)
* Expect devs to change this in multi-theme

Reply:
This is not at all a PR blocker anyway - feel free to ignore and we can address when an actual example of it causing a problem appears.

Reply:
Added a note to the README. I think there's easy ways to handle this, but I think that should be a discussion within the project that _wants_ to use Preprocessor and multi-design. But really good to think about early and often!